zilliqa/src/pool.rs

```diff
@@ -151,15 +151,14 @@ impl TransactionPool {
                 // to broadcast a new transaction to the network.
                 if ReadyItem::from(existing_txn) >= ReadyItem::from(&txn) {
                     return false;
-                } else {
-                    // Remove the existing transaction from `hash_to_index` if we're about to replace it.
-                    self.hash_to_index.remove(&existing_txn.hash);
                 }
+                // Remove the existing transaction from `hash_to_index` if we're about to replace it.
+                self.hash_to_index.remove(&existing_txn.hash);
             }
         } else {
             // now we've confirmed it's nonceless, so ensure its index isn't duplicate
             if let Some(existing_nonceless_txn) = self.transactions.get(&txn.mempool_index()) {
-                warn!(tx = ?existing_nonceless_txn, "Duplicate-indexed nonceless transactions encountered, this shouldn't happen. Ignoring the new one.");
+                info!(tx = ?existing_nonceless_txn, "Duplicate-indexed nonceless transactions encountered, ignoring the new one.");
```

Comment:
This turns out to actually happen normally when `consensus.new_transaction()` is called for every transaction inside `consensus.execute_block()`. Thus downgraded to an `info()` diagnostic. The updated `mark_executed()` ensures these transactions still don't get double-executed, so it's not a problem.

Reply:
Can we simplify lines 143-147 in this function by using `.mempool_index()`.

i.e.
```rust
if let Some(existing_txn) = self.transactions.get(txn.mempool_index()) {
  ...
}
```
should cover both cases?

Reply:
Good call. This has to be a no-op for the nonceless case, and refactoring it that way _should_ keep that property - I've added a comment explaining why, and the tests are passing, so should be good and does make it more concise and readable.